test/core/event_engine/default_engine_methods.cc

```diff
@@ -0,0 +1,107 @@
+// Copyright 2022 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include <grpc/support/port_platform.h>
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+
+#include <grpc/event_engine/event_engine.h>
+#include <grpc/grpc.h>
+
+#include "src/core/lib/event_engine/event_engine_factory.h"
+#include "src/core/lib/gprpp/sync.h"
+#include "test/core/util/test_config.h"
+
+namespace {
+
+using ::grpc_event_engine::experimental::EventEngine;
+using ::grpc_event_engine::experimental::GetDefaultEventEngine;
+
+class DefaultEngineTest : public testing::Test {
+ protected:
+  // Does nothing, fills space that a nullptr could not
+  class FakeEventEngine : public EventEngine {
+   public:
+    FakeEventEngine() = default;
+    ~FakeEventEngine() override = default;
+    absl::StatusOr<std::unique_ptr<Listener>> CreateListener(
+        Listener::AcceptCallback /* on_accept */,
+        absl::AnyInvocable<void(absl::Status)> /* on_shutdown */,
+        const grpc_event_engine::experimental::EndpointConfig& /* config */,
+        std::unique_ptr<
+            grpc_event_engine::experimental::
+                MemoryAllocatorFactory> /* memory_allocator_factory */)
+        override {
+      return absl::UnimplementedError("test");
+    };
+    ConnectionHandle Connect(
+        OnConnectCallback /* on_connect */, const ResolvedAddress& /* addr */,
+        const grpc_event_engine::experimental::EndpointConfig& /* args */,
+        grpc_event_engine::experimental::MemoryAllocator /* memory_allocator */,
+        Duration /* timeout */) override {
+      return {-1, -1};
+    };
+    bool CancelConnect(ConnectionHandle handle) override { return false; };
+    bool IsWorkerThread() override { return false; };
+    std::unique_ptr<DNSResolver> GetDNSResolver(
+        const DNSResolver::ResolverOptions& options) override {
+      return nullptr;
+    };
+    void Run(Closure* closure) override{};
+    void Run(absl::AnyInvocable<void()> closure) override{};
+    TaskHandle RunAfter(Duration when, Closure* closure) override {
+      return {-1, -1};
+    }
+    TaskHandle RunAfter(Duration when,
+                        absl::AnyInvocable<void()> closure) override {
+      return {-1, -1};
+    }
+    bool Cancel(TaskHandle handle) override { return false; };
+  };
+};
+
+TEST_F(DefaultEngineTest, SharedPtrGlobalEventEngineLifetimesAreValid) {
```

Comment:
let's add a test with a dozen-ish threads doing `GetDefaultEventEngine().reset();` in a loop for a few seconds

Reply:
Added, and it found a race. I reworked the default engine storage using weak_ptrs, it's a bit cleaner now.